### PR TITLE
Separate the legacy middleware from the GQL API middleware

### LINKED ISSUE #4042: Disable unneeded middlewares in the GraphQL API
Following #3631, we would like to disable any unneeded middleware. Some middlewares are actually needed sometimes but not everytime though, those should become lazy middlewares.

### PR DESCRIPTION
**Pull request by NyanKiyoshi**
Closes #4042.

- Disables unneeded middleware on the GQL API endpoint and for the 1.0 views;
- Makes heavy request attributes lazy.

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] The code is documented (docstrings, project documentation).
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

**Comment by NyanKiyoshi** [comment]
The coverage loss is coming from disabling the Google Analytics middleware by default (when the API key is not provided). As it performs quite a few calls to get a value that end up not never being used (checked it).

File: saleor/settings.py [hunk 1]
```diff
@@ -194,24 +194,64 @@ def get_bool_from_env(name, default_value):
...
+# Only used to dispatch requests to correct middlewares, please don't directly use it.
+MIDDLEWARE = ["saleor.core.middleware.middleware_dispatcher"]
+
+# Middleware for Saleor 1.0 (storefront and dashboard)
+DJANGO_MIDDLEWARE = [
+    # SessionMiddleware stores and manages session data from/to a cookie.
     "django.contrib.sessions.middleware.SessionMiddleware",
+    # SecurityMiddleware redirect the users to HTTPS when required/requested.
     "django.middleware.security.SecurityMiddleware",
+    # CommonMiddleware has for job to:
```
Comment (maarcingebala) [inline saleor/settings.py]: I think most of these comments are unnecessary, they're documenting parts of Django or third-party libraries and should be already documented their docs. When it comes to comments related to our middlewares, they should be added as docstrings inside middlewares' definitions.

File: saleor/core/middleware.py [hunk 2]
```diff
@@ -38,10 +104,13 @@ def discounts(get_response):
     """Assign active discounts to `request.discounts`."""
 
     def middleware(request):
-        discounts = Sale.objects.active(date.today()).prefetch_related(
-            "products", "categories", "collections"
+        request.discounts = SimpleLazyObject(
```
Comment (patrys) [inline saleor/core/middleware.py]: Since that's a queryset, what do we gain from making it even lazier?

File: saleor/core/middleware.py [hunk 3]
```diff
@@ -15,9 +28,62 @@
...
 
 
+def django_only_middleware(middleware):
+    @wraps(middleware)
+    def wrapped(get_response):
+        api_path = reverse("api")
+        handler = middleware(get_response)
+
+        @wraps(handler)
+        def conditional(request):
```
Comment (patrys) [inline saleor/core/middleware.py]: Can we split this to a separate decorator and return `something(handler)`?

**Pull request merged**